README.md
# miditk-smf

A Python toolkit for working with Standard MIDI files

## Quickstart

Install:

```console
pip install miditk-smf
```

Usage:

```py
from miditk.smf import MidiFileWriter

# Open file for writing in binary mode.
with open('minimal.mid', 'wb') as smf:
    # Create standard MIDI file writer.
    midi = MidiFileWriter(smf)

    # Write file and track header for Type 0 file, one track, 96 pulses per
    # quarter note (ppqn). These are also the default parameter values.
    midi.header(format=0, num_tracks=1, tick_division=96)
    midi.start_of_track()

    # Set tempo to 120 bpm in µsec per quarter note.
    # When no tempo is set in the MIDI file, sequencers will generally assume
    # it to be 120 bpm.
    midi.tempo(int(60_000_000 / 120))

    # Add MIDI events.
    midi.note_on(channel=0, note=0x40)
    # Advance 192 ticks (i.e. a half note).
    midi.update_ticks(192)
    midi.note_off(channel=0, note=0x40)

    # End track and midi file.
    midi.end_of_track()
    midi.eof()
```

For more examples, see the [Usage examples](#usage-examples) section below.


## Overview

`miditk-smf` is a general-purpose library for the parsing and generation of
Standard MIDI Files (SMF). The package is part of several planned packages
under the common top-level package namespace `miditk`. This package mainly
provides the `miditk.smf` sub-package for handling standard MIDI files.
Additional sub-packages with more specialised MIDI libraries and tools may be
developed and distributed as separate package distributions in the future.


### Compatibility

`miditk-smf` works with (C)Python >= 3.8 and PyPy 3.


## Installation

`miditk-smf` is installable via [pip](https://pypi.org/project/pip/)
from the [Python Package Index](https://pypi.org/project/miditk-smf/):

```console
pip install miditk-smf
```

It is provided as a source distribution and a universal Python wheel for all
supported Python versions and operating systems. It only depends on the Python
standard library.


## Package contents

`miditk.common`:

A collection of constants from the MIDI specification used by sub-packages and
general data types for working with MIDI events.

`miditk.smf`:

An event-based standard MIDI file (SMF) parsing and generation framework.

`miditk.smf.api`:

Base event handler classes, which can be subclassed for specialised event
handling.

`miditk.smf.converters`:

A collection of functions that converts the special data types used in midi
files to and from byte strings.

`miditk.smf.parser`:

The main binary MIDI file data parser.

`miditk.smf.reader`:

Combines the parser with an event handler class.

`miditk.smf.sequence`:

An event handler, which stores all MIDI events from a MIDI file in a
`MidiSequence` container class.

`miditk.smf.writer`:

An event handler to write out MIDI events to a standard MIDI File.


## Usage examples

The following section contains a few code examples, which demonstrate several
usage scenarios for the different modules in the package. For more examples see
also the scripts in the [examples] directory of the source distribution.


### Parsing a standard MIDI file

The `miditk.smf` module provides the `MidiSequence` container class, which uses
its own MIDI event handler class to collect all information and events from
parsing a midi file. Use the `MidiSequence.fromfile()` class method to parse a
standard MIDI file.

You can then use several convenience methods of the returned `MidiSequence`
instance to access information about the midi file properties or events.

```py
from miditk.smf import MidiSequence

# Do parsing
sequence = MidiSequence.fromfile(sys.argv[1])

# Print some info from the MIDI file header,
# e.g. number of tracks, events sequence name.
print(sequence)
# Print a list of events with event type, data and timestamp.
sequence.dump_events()

# Iterate over all sysex events in track 0.
# If track is not specified, sysex_events() yields all sysex events
# in all tracks.
for ev in sequence.sysex_events(track=0):
    print("Sysex event ({} bytes) @ {:.2f}".format(len(ev.data), ev.timestamp))

# Iterate over all events sorted by timestamp and then track.
for time, group in sequence.events_by_time():
    for ev in group:
        handle_event(ev)
```


### Changing MIDI events in-stream

The event-based parsing allows to handle MIDI events as they are read (or
received via MIDI in). You need to define a sub-class of
`miditk.smf.BaseMidiEventHandler` or `miditk.smf.NullMidiEventHandler` and
overwrite only the event handling methods for the events you are interested in.

The following example transposes all note on/off events by an octave (i.e. 12
semitones):

```py
import sys
from miditk.smf import MidiFileReader, MidiFileWriter

# MidiFileWriter is a sub-class of NullMidiEventHandler.
class Transposer(MidiFileWriter):
    """Transpose note values of all note on/off events by 1 octave."""

    def note_on(self, channel, note, velocity):
        super().note_on(self, channel, min(127, note + 12), velocity)

    def note_off(self, channel, note, velocity):
        super().note_off(self, channel, min(127, note + 12), velocity)

infile = sys.argv.pop(1)
outfile = sys.argv.pop(1)

# Create the parser and event handler
with open(outfile, 'wb') as smf:
    midiout = Transposer(smf)
    midiin = MidiFileReader(infile, midiout)

    # Now do the processing.
    midiin.read()
```


## Development

Clone the Git repository:

```console
git clone https://github.com/SpotlightKid/miditk-smf.git
cd miditk-smf
```

Install tox:

```console
pip install tox
```

Or via your Linux distribution package manager, e.g. on debian/Ubuntu:

```console
sudo apt-get install python-tox
```

Or on Arch Linux:

```console
sudo pacman -S python-tox
```

Run the tests via tox for all Python versions configured in `tox.ini`:

```console
tox
```

If all is well, create a new git branch and start hacking and then contribute
your changes by opening a [pull
request](https://github.com/SpotlightKid/miditk-smf/pulls) on GitHub.


## Code QA

The included Makefile is set up to run several Python static code checking and
reporting tools. To print a list of available Makefile targets and the tools
they run, simple run:

```console
make
```

Then run the Makefile target of your choice, e.g.:

```console
make flake8
```

Unless noted otherwise, these targets run all tools directly, i.e. without tox,
which means they need to be installed in your Python environment. You can use
[hatch] to create a virtual environments for general development tasks or for
specific tasks as, for example, building the documentation. Dependencies and
tools needed for these tasks will be installed automatically into these
environments on creation:

To show which special environments are defined:

```console
hatch env show
```

To create and enter e.g. the "dev" environment:

```console
hatch --env dev shell
```


## Documentation

Package documentation is generated by Sphinx. The documentation can be build
with:

```console
make docs
```

After a successful build the documentation index is opened in your web
browser.


## Authors and License

The `miditk` package is written by Christopher Arndt and licensed under the MIT
License.

The the structure of the `miditk.smf` sub-package owes inspiration to the
[Python Midi] [^1] package, written by <maxm@maxm.dk>.


[examples]: https://github.com/SpotlightKid/miditk-smf/tree/master/examples
[python midi]: http://web.archive.org/web/20100919095628/http://www.mxm.dk/products/public/pythonmidi/
[^1]: Original web site, now defunct: http://www.mxm.dk/products/public/pythonmidi/


CHANGELOG.md
# Changelog

## 0.3.1 (2023-08-21)

Fixes:

-   Fixed writing polyphonic pressure events in `MidiFileWriter` (#16).

    Thanks to Jared Boone (@jboone) for the fix.

Project updates:

-   Oldest supported Python version is now 3.8.


## 0.3.0 (2022-12-08)

Project updates:

-   Switched from `setuptools` to PEP-517-compatible build setup using
    `hatchling`.
-   Removed Python 2 support (now requires Python >= 3.7).
-   Removed `six` dependency.
-   Removed requirements files.
-   Converted README, LICENSE and CHANGELOG to Markdown format.
-   Added automatic testing and deployment of source distribution and
    wheel to PyPI via GitHub actions.

Changes:

-   Made `track` argument to `BaseMidiEventHandler.start_of_track` optional.
-   Split `MidiFileWriter` into base class and event handler sub-class.
    Base class handles event serialization and output.
-   `MidiSequence` instances now save tempo and time/key signature events.
-   `MidiSequence.fromfile` can now accept a custom handler class/instance.
-   Added command-line script `miditk-mid2syx`.
-   Updated examples and removed Python 2 compatibility idioms.

Fixes:

-   Fixed writing sysex message events in `MidiFileWriter` (#7).
-   Removed unused `dispatch_controllers` attribute in `BaseMidiEventHandler`.
-   Fixed `MidiSequence` example in readme and added example script.
-   Minor fixes: removed unused import; added geany project file to git ignores;
    fixed code formatting issues.
-   Reformated code with `black` and `isort`.

## 0.2.2 (2019-01-04)

-   Added example script to convert a type 1 MIDI file into a type 0
    format file.
-   Moved previously private `MidiEvent` class to `miditk.smf.api`
    module and exposed it in `miditk.smf` module name space and improved
    and fixed several aspects of the `MidiEvent` class.
-   Fixed parsing of polyphonic pressure events.
-   Fixed output for polyphonic pressure events in
    `PrintingMidiEventHandler`.
-   Fix time keeping in `ObjectMidiEventHandler` for multi-track files:
    reset `current_time` attribute when new track starts.
-   Fixed: `BaseMidiEventHandler` used wrong handler method name for
    sysex messages.
-   Updated base/dev requirements.

## 0.2.1 (2018-08-20)

-   Fixed missing string formatting args in
    `PrintingMidiEventHandler.key_signature` method.
-   Added `example_program_change_map.py` script to examples.
-   Ensure support for Python 3.7.

## 0.2.0 (2018-04-18)

-   Removed Cython version of `miditk.smf.converters` module
    (`miditk.smf._speedups`). `miditk-smf` now only requires `six`.
-   Fixed `tox.ini` to work on Windows too.
-   Improved and updated readme file.

## 0.1.0 (2018-04-09)

Initial release


miditk/smf/sequence.py
#
# miditk/smf/sequence.py
#
"""A MIDI stream event handler class to convert an SMF file into a container
object for a sequence of MIDI time-stamped events."""

# standard library imports
import logging
from inspect import isclass
from itertools import groupby
from operator import attrgetter

# package-specific imports
from ..common.constants import (
    END_OF_EXCLUSIVE,
    END_OF_TRACK,
    KEY_SIGNATURE,
    SEQUENCE_NAME,
    SYSTEM_EXCLUSIVE,
    TEMPO,
    TIME_SIGNATURE,
)
from .api import BaseMidiEventHandler
from .converters import tointseq
from .reader import MidiFileReader

__all__ = ("MidiSequence", "ObjectMidiEventHandler")
log = logging.getLogger(__name__)


class MidiSequence(list):
    """Container for a sequence of time-stamped MIDI events."""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.tracks = {}

    def __repr__(self):
        title = getattr(self, "sequence_name", "")
        return "<MidiSequence tracks=%i events=%i, title='%s'>" % (
            self.num_tracks,
            len(self),
            title,
        )

    def dump_events(self):
        for i, ev in enumerate(sorted(self, key=attrgetter("time", "track"))):
            if ev.channel is not None:
                print(
                    "%03i: Channel event - type: %02XH, channel: %X, data: %r [%.2f]"
                    % (i, ev.type, ev.channel, tointseq(ev.data), ev.time)
                )
            elif ev.meta_type:
                print(
                    "%03i: Meta event - type: %02XH, data: %r [%.2f]"
                    % (i, ev.meta_type, tointseq(ev.data), ev.time)
                )
            elif ev.type == SYSTEM_EXCLUSIVE:
                print("%03i: Sysex_event - size: %i [%.2f]" % (i, len(ev.data) + 1, ev.time))

    def events_by_ticks(self):
        return groupby(sorted(self, key=attrgetter("ticks", "track")), key=attrgetter("ticks"))

    def events_by_time(self):
        return groupby(sorted(self, key=attrgetter("time", "track")), key=attrgetter("time"))

    def events_by_track(self):
        return groupby(sorted(self, key=attrgetter("track", "time")), key=attrgetter("track"))

    def channel_message_events(self, track=None):
        for event in sorted(self, key=attrgetter("track", "time")):
            if event.channel and (track is None or event.track == track):
                yield event

    def meta_events(self, track=None):
        for event in sorted(self, key=attrgetter("track", "time")):
            if event.meta_type and (track is None or event.track == track):
                yield event

    def sysex_events(self, track=None):
        for event in sorted(self, key=attrgetter("time", "track")):
            if event.type == SYSTEM_EXCLUSIVE and (track is None or event.track == track):
                yield event

    @classmethod
    def fromfile(cls, infile, handler=None, *args, **kwargs):
        if handler is None:
            handler = ObjectMidiEventHandler

        sequence = cls()

        if isclass(handler):
            handler = handler(sequence, *args, **kwargs)

        midiin = MidiFileReader(infile, handler)
        midiin.read()
        return sequence


class ObjectMidiEventHandler(BaseMidiEventHandler):
    """Collects all information from a MIDI file in a container object.

    On class instantiation you need to pass an instance to the constructor,
    which supports the Python list interface and attribute assignment. If
    None is passed (the default), a ``MidiSequence`` instance will be created
    and used instead.

    """

    meta_events = [
        "copyright",
        "cuepoint",
        "device_name",
        "end_of_track",
        "instrument_name",
        "key_signature",
        "lyric",
        "marker",
        "midi_ch_prefix",
        "midi_port",
        "program_name",
        "sequence_name",
        "sequence_number",
        "sequencer_specific",
        "smtp_offset",
        "tempo",
        "text",
        "time_signature",
    ]
    channel_events = [
        "aftertouch",
        "channel_pressure",
        "controller_change",
        "note_off",
        "note_on",
        "pitch_bend",
        "program_change",
    ]

    def __init__(self, instance=None, debug=False, encoding="utf-8"):
        if instance is None:
            instance = MidiSequence()

        super().__init__(encoding=encoding)
        instance.tempo_map = []
        instance.time_signature_map = []
        instance.key_signature_map = []
        self._instance = instance
        self.current_time = 0.0
        self.current_tempo = 500000
        self.current_time_signature = None
        self.current_key_signature = None
        self.tick_division = 96
        self.running_status = None
        self._sysex_continuation = False
        self._sysex_buffer = None
        self.debug = debug

    def __getattr__(self, name):
        if name in self.meta_events:
            return self.meta_event
        elif name in self.channel_events:
            return self.channel_message
        else:
            raise AttributeError(name)

    #########################
    ## time handling event handlers.
    ## They should be overwritten with care

    def update_ticks(self, ticks=0, relative=True):
        """Update the global time."""
        BaseMidiEventHandler.update_ticks(self, ticks, relative)
        self.current_time += ticks * (self.current_tempo / 1000.0 / self._instance.tick_division)

        if self.debug:
            log.debug("Timer update: %s ticks", ticks)

    def add_event(self, event, track=None):
        """Add given event to the container object.

        Takes care of adding the current sequence time to the object.

        """
        event.time = self.current_time
        event.ticks = self.absolute_time
        event.relative_time = self.relative_time
        event.track = self.current_track if track is None else track
        self._instance.append(event)

    #########################
    ## Parsing (non-midi) events

    ## MIDI file header handler

    def header(
        self,
        format=0,
        num_tracks=1,
        tick_division=96,
        metrical=True,
        fps=0xE7,
        frame_resolution=0x28,
    ):
        """Handle MIDI file header."""
        for key, value in locals().items():
            if key != "self" and not key.startswith("_"):
                setattr(self._instance, key, value)

        if metrical:
            log.debug(
                "MIDI file format: %s, no. of tracks: %i, " "tick division: %i ppqn",
                format,
                num_tracks,
                tick_division,
            )
        else:
            log.debug(
                "MIDI file format: %s, no. of tracks: %i, fps: %i, " "resolution: %i",
                format,
                num_tracks,
                fps,
                frame_resolution,
            )

    def eof(self):
        """Handle end of MIDI file."""
        self.current_track = None

        if self.debug:
            log.debug("End of file reached.")

    def start_of_track(self, track):
        """Handle start of new track."""
        if self.debug:
            log.debug("Start of track #%i.", track)

        self.current_track = track
        self.current_time = 0.0

    #############################
    ## Channel events

    def channel_message_event(self, event, *args):
        """Handle channel messages events."""
        if self.debug:
            log.debug("Channel message event: %s", event)

        self.add_event(event)

    ###############
    ## sysex event

    def invalid_event(self, event):
        """Handle invalid events."""
        log.warning("Received invalid event type (%X) of %i bytes.", event.type, len(event.data))

    def sysex_event(self, event):
        """Handle system exclusive message events.

        ..note:
            Some MIDI files contain sysex messages which are distributed over
            several events, where the event containing the first part of the
            message has the normal 0xF0 status but the following events have
            a 0xF7 (normally used for escape sequence) and no start-of-sysex,
            i.e. all data bytes. The last part of the message in the sequence
            will have end-of-system-exclusive 0xF7 as the last data byte.

            We assemble these sysex messages here into a single sysex event
            with the timestamp of the event with the first part of the message.

        """
        if self.debug:
            log.debug("System exclusive event, %i bytes", len(event.data) + 1)

        if tointseq(event.data)[-1] == END_OF_EXCLUSIVE:
            if self._sysex_continuation:
                if self.debug:
                    log.debug("Sysex continuation in effect. Appending event data.")

                self._sysex_buffer.data += event.data
                event, self._sysex_buffer = self._sysex_buffer, None

            if self.debug:
                log.debug("Storing sysex mssage of %i bytes", len(event.data) + 1)

            self.add_event(event)
            self._sysex_continuation = False
        else:
            if self.debug:
                log.debug("Sysex message not terminated. Assuming continuation.")

            if self._sysex_continuation:
                self._sysex_buffer.data += event.data
            else:
                self._sysex_buffer = event

            self._sysex_continuation = True

    def escape_sequence(self, event):
        if self.debug:
            log.debug("Escape sequence, value: %r" % event.data)

        self.add_event(event)

    #####################
    ## meta events

    def meta_event(self, event, *args):
        """Dispatch meta events."""
        # TEMPO = 0x51 (51 03 tt tt tt (tempo in us/quarternote))
        track = self.current_track

        if event.meta_type == TEMPO:
            b1, b2, b3 = tointseq(event.data)
            event.tempo = (b1 << 16) + (b2 << 8) + b3
            # uses 3 bytes to represent time between quarter
            # notes in microseconds
            self.tempo(event.tempo)

        # TIME_SIGNATURE = 0x58 (58 04 nn dd cc bb)
        elif event.meta_type == TIME_SIGNATURE:
            nn, dd, cc, bb = tointseq(event.data)
            self.time_signature(nn, dd, cc, bb)

        # KEY_SIGNATURE = 0x59 (59 02 sf mi)
        elif event.meta_type == KEY_SIGNATURE:
            sf, mi = tointseq(event.data)
            self.key_signature(sf, mi)

        # SEQUENCE_NAME = 0x03 (03 len text...)
        elif event.meta_type == SEQUENCE_NAME:
            self.sequence_name(event.data)

        # END_OF_TRACK = 0x2F
        elif event.meta_type == END_OF_TRACK:
            self.end_of_track(track)
        else:
            if self.debug:
                log.debug("Meta event: %s", event)

        self.add_event(event, track)

    def end_of_track(self, track=None):
        """Handle end of current track."""
        self.current_track = None

        if self.debug:
            log.debug("End of track #%i.", track)

    def sequence_name(self, name):
        """Handle Sequence / Track name events."""
        name = name.decode("latin1", "replace")
        if self.current_track == 0:
            if self.debug:
                log.debug("Sequence name: %s", name)

            self._instance.sequence_name = name
        else:
            if self.debug:
                log.debug("Track name (%02i): %s", self.current_track, name)

            self._instance.tracks[self.current_track] = name

    def tempo(self, value):
        """Handle tempo change events."""
        if not hasattr(self._instance, "initial_tempo"):
            self._instance.initial_tempo = value

        self.current_tempo = value
        bpm = 60000000.0 / value
        ms_per_quarter = value / 1000.0
        self._instance.tempo_map.append((self.current_time, (bpm, ms_per_quarter)))

        if self.debug:
            log.debug("Tempo: %.2f ms/quarter note (%.2f BPM)", ms_per_quarter, bpm)

    def time_signature(self, nn, dd, cc, bb):
        """Handle time signature change events."""
        self.current_time_signature = (nn, dd)
        self._instance.time_signature_map.append((self.current_time, (nn, dd)))

        if self.debug:
            log.debug("Time signature: %i/%i (%i cpm, %i dpqn)", nn, dd**2, cc, bb)

    def key_signature(self, sf, mi):
        """Handle key signature change events."""
        self.current_key_signature = (sf, mi)
        self._instance.key_signature_map.append((self.current_time, (sf, mi)))

        if self.debug:
            log.debug("Key signature: %s %s", sf, mi)


if __name__ == "__main__":
    import sys

    if len(sys.argv) > 1:
        test_file = sys.argv[1]
    else:
        print("Usage: sequence.py MIDIFILE")

    logging.basicConfig(level=logging.INFO)
    # do parsing
    sequence = MidiSequence.fromfile(test_file)
    print(sequence)
    sequence.dump_events()
